blog/2020-07-13-alpine-python.md

```diff
@@ -0,0 +1,156 @@
+
+But there is a game-changer which makes the musl one more useful as the glibc based runtime. It is the memory allocator [jemalloc](http://jemalloc.net/), a general-purpose malloc implementation that emphasizes fragmentation avoidance and scalable concurrency support. There is an interesting effect which I found on some blogpost about Rust. There were some developers who saw that musl is much faster as glibc with jemalloc, but glibc is slower with jemalloc as without. For sure, the benefit with glibc and jemalloc is not the speed as they optimize memory management, but musl get both benefits. While the difference between pure musl and glibc can be ignored, the difference between musl + jemalloc and glibc are substantial (with disabled GCC memory allocator built-in optimization). Yes, jemalloc is compatible today with musl, and there was a time which it was not.
+
+## Compiler
+
+How do you compile Python is also essential. There were at least statements from Fedora or Redhat about disable `semantic-interposition` to become a high-performance boost. I was not able to reproduce this on GCC 9.3.0, but I also saw no adverse side effects. I can recommend disabling the semantics like the built-in allocator optimization and link jemalloc at build time. I will also recommend using the `-O3` optimization. We never saw an issue with this aggressive optimization on our target platforms. I need to say, unlike the distro Python runtime interpreters, we don't need to run everywhere. So we can use the `--enable-optimizations` without any overwrite and add more flags. I can say today, PGO/LTO/O3 make Python faster and at least work on our target CPUs.
+
+## Python packages
+
+Yes, Alpine indeed has no manylinux compatibility with musl. If you don't cache your builds, it needs every time compiling the C extensions on that packages. This process would take time. More time if you also crossbuild with Qemu for different CPU architectures. You can not get precompiled binaries from PyPi. I don't see a loose there because the packages are mostly also not optimized for the target system. We created our wheel [index](https://wheels.home-assistant.io/) and [backend]( https://github.com/home-assistant/wheels) to compile and hold it up to date over CI agents. For sure, not every project can take this effort, but also not all repository haver over 1k packages for each CPU architecture, and the build time of the docker file is not so important at all.
+
+## Alpine Linux
+
+It is a great base system for Container and allows us to provide the best experience to our User. A big thanks to Alpine Linux, musl, and jemalloc, which make that possible.
+
+Here is a Benchmark between the distro Python and our optimization (GCC 9.3.0 / musl / Python 3.8.3)
```

Comment:
```suggestion
The table shows the results between the Alpine Linux's Python runtime and our optimization (GCC 9.3.0/musl/Python 3.8.3).
```

IMHO, there should be a little more context here (how the results were gathered, how it can be reproduced, how the selection was made, etc.). The naked table is hard to interpret, especially as at least one run is slower.

Reply:
We can also remove it 🤷‍♂️